### PR TITLE
[wol]  Extend wol to support sending magic pattern in udp payload

### PR DESCRIPTION
**Pull request by w1nda**
#### Why I did it
Although the magic pattern is fixed, we can send the magic pattern in UDP payload to leverage the benefits of UDP.

##### Work item tracking
- Microsoft ADO **(number only)**:

#### How I did it
Add code to support sending payload in udp and provide CLI parameters for it.

#### How to verify it
Unit test and run tests in sonic-mgmt

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

File: src/sonic-nettools/wol/src/wol.rs [hunk 1]
```diff
@@ -30,29 +46,31 @@ struct WolArgs {
...
 
+    /// The flag to indicate if send udp packet [default: false]
+    #[arg(short, long, default_value_t = false, conflicts_with("broadcast"))]
+    udp: bool,
+
+    /// The destination ip address, both IPv4 address and IPv6 address are supported [default: 255.255.255.255]
+    #[arg(short = 'a', long, default_value_t = String::from("255.255.255.255"), requires_if(ArgPredicate::IsPresent, "udp"))]
+    ip_address: String,
+
+    /// The the destination udp port. [default: 9]
```
Comment (w1nda) [inline src/sonic-nettools/wol/src/wol.rs:60]: TODO: remove duplicated hint.

image

File: src/sonic-nettools/wol/src/socket.rs [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+            let _in = unsafe {
+                std::ffi::CStr::from_ptr(addr_ref.ifa_name).to_str().unwrap()
+            };
+            if _in == intf_name {
+                addr_ll.sll_ifindex = unsafe {
+                    libc::if_nametoindex(addr_ref.ifa_name) as i32
+                };
+                addr_ll.sll_addr = unsafe { (*(addr_ref.ifa_addr as *const libc::sockaddr_ll)).sll_addr };
+                break;
```
Comment (w1nda) [inline src/sonic-nettools/wol/src/socket.rs:103]: TODO: Raise error when there is no interface found.

File: src/sonic-nettools/wol/src/wol.rs [hunk 3]
```diff
@@ -83,7 +101,7 @@ impl std::fmt::Display for WolErr {
     }
 }
 
-enum WolErrCode {
+pub enum WolErrCode {
     SocketError = 1,
     InvalidArguments = 2,
     UnknownError = 999,
```
[the remark this answers is not in the document]
Comment (w1nda) [reply src/sonic-nettools/wol/src/wol.rs:108]: Nice catch, I will update it.

File: src/sonic-nettools/wol/src/wol.rs [hunk 4]
```diff
@@ -679,9 +693,40 @@ mod tests {
...
+            "error: the following required arguments were not provided:\n  --udp\n\nUsage: wol --udp --udp-port <UDP_PORT> <INTERFACE> <TARGET_MAC>\n\nFor more information, try '--help'.\n"
+        );
+        // Ip address should be specified with udp flag
+        let result = WolArgs::try_parse_from(&["wol", "eth0", "00:01:02:03:04:05", "-a", "192.168.1.1"]);
+        assert!(result.is_err());
+        assert_eq!(
+            result.unwrap_err().to_string(),
+            "error: the following required arguments were not provided:\n  --udp\n\nUsage: wol --udp --ip-address <IP_ADDRESS> <INTERFACE> <TARGET_MAC>\n\nFor more information, try '--help'.\n"
+        );
+
```
[the remark this answers is not in the document]
Comment (w1nda) [reply src/sonic-nettools/wol/src/wol.rs:730]: Will do

File: src/sonic-nettools/wol/src/wol.rs [hunk 5]
```diff
@@ -1,15 +1,26 @@
...
 use std::str::FromStr;
+use std::sync::Mutex;
 use std::thread;
 use std::time::Duration;
 
+use crate::socket::{WolSocket, RawSocket, UdpSocket};
+
 const BROADCAST_MAC: [u8; 6] = [0xff, 0xff, 0xff, 0xff, 0xff, 0xff];
 
+pub static VERBOSE_OUTPUT: Mutex<bool> = Mutex::new(false);
```
[the remark this answers is not in the document]
Comment (w1nda) [reply src/sonic-nettools/wol/src/wol.rs:16]: Using mutex is safer although there is no race condition may happen

**Review by lizhijianrd — APPROVED** [review]

**Comment by w1nda** [comment]
Hi @yxieca, cloud you help merge this PR? thanks

**Comment by w1nda** [comment]
This cherry-pick depends on PR [#21136](https://github.com/sonic-net/sonic-buildimage/pull/21136)
which is merged, 
So, I try remove the conflict tag and add approval tag to retry

**Pull request merged**